#Python Language  + Data Gathering

## Objectives

* Retrieve data from a public API
* Manipulate nested data structures: JSON
* Use python lists and dictionaries
* Make python functions



## API: Application programming interface

_"a set of clearly defined methods of communication between various software components."_

In [12]:
%%html
<iframe width="560" height="315" src="https://www.youtube.com/embed/s7wmiS2mSXY" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Example of different parts to define an API:
* A protocol (eg: `https`)
* A server (eg: `httpbin.org`)
* A method name / location (eg: `/get`)
* A set of arguments (eg: `hello=world` and `foo=bar`)

➥ `https://httpbin.org/get?hello=world&foo=bar`

## Villo API

We will be using an API that provides information about the Villo stations and bike availability in Brussels. The API is documented on the [OpenDataSoft](https://public.opendatasoft.com/explore/dataset/jcdecaux-bike-stations-data-rt/information/) website.

![villo_photo](http://2.bp.blogspot.com/_EedX2cyld30/TTbxocuUIGI/AAAAAAAAAcY/v_uETl4Vntg/s1600/IMG_3964.JPG)

In [13]:
%%html
<iframe src="https://public.opendatasoft.com/explore/dataset/jcdecaux-bike-stations-data-rt/information/" style="width: 100%; height: 800px"></iframe>

## Let's retrive the information from the Villo API in this notebook

The library [requests](http://docs.python-requests.org/en/master/) makes retrieving data in Python pretty easy. We will start by defining the url of the API:

In [14]:
URL_Villo = "https://public.opendatasoft.com/api/records/1.0/search/?dataset=jcdecaux-bike-stations-data-rt&q=&facet=banking&facet=name&exclude.name=BORNE+TEST+ALLEE+VERTE&exclude.name=BORNE+ATELIER+MECANOS+SPL&facet=bonus&facet=status&facet=contract_name&refine.contract_name=Bruxelles-Capitale&rows=50"

Note: You may have to use a different value for `URL_Villo` if the
default arguments are not sufficient for your query.

In [15]:
import requests
response = requests.get(URL_Villo)
display(response)

<Response [200]>

## What is this &lt;Response 200&gt; ?

Have a look at the [HTTP Status Codes](http://www.restapitutorial.com/httpstatuscodes.html) to understand these _Error 404_ or _Error 500_ you may have seen on the web.

In [16]:
response.status_code

200

In [17]:
# Show the error message if the retrieval failed for some reason:
if response.status_code == 200:
    print("The request went fine.")
else:
    print('There was a problem with the request')

The request went fine.


Let's now see what is inside the response we got:

In [18]:
response.content

b'{"nhits": 351, "parameters": {"dataset": "jcdecaux-bike-stations-data-rt", "rows": 50, "start": 0, "facet": ["banking", "name", "bonus", "status", "contract_name"], "format": "json", "timezone": "UTC"}, "records": [{"datasetid": "jcdecaux-bike-stations-data-rt", "recordid": "6e8b2f6a144a66d65da8d2974bd1a4db0eee4841", "fields": {"available_bike_stands": 12, "bike_stands": 25, "number": 305, "address": "PETER BENOIT - PLACE PETER BENOIT PLAATS 2-4", "name": "305 - PETER BENOIT", "bonus": "False", "banking": "False", "contract_name": "Bruxelles-Capitale", "status": "OPEN", "available_bikes": 12, "position": [50.895592, 4.387965], "last_update": "2024-02-26T12:45:24+00:00"}, "geometry": {"type": "Point", "coordinates": [4.387965, 50.895592]}}, {"datasetid": "jcdecaux-bike-stations-data-rt", "recordid": "39042b9903ec9a87d9fd670c5999fb52ad8d9e6d", "fields": {"available_bike_stands": 11, "bike_stands": 25, "number": 200, "address": "VETERANS COLONIAUX / KOLONIALE VETERANEN - BD JULES GRAIND

When looking at the content of the response, we can see it that it looks like Python dictionaries and lists, and it is probably encoded in [JSON (JavaScript Object Notation)](https://www.json.org/). The other popular formats for data structure are XML (mainly old stuff) and CSV (spreadsheet compatible).

Let's try parsing the data as if it were JSON, in order to get it as imbricated dictionaries and lists.  `requests` contains a shortcut do do this.

In [19]:
data = response.json()
data

{'nhits': 351,
 'parameters': {'dataset': 'jcdecaux-bike-stations-data-rt',
  'rows': 50,
  'start': 0,
  'facet': ['banking', 'name', 'bonus', 'status', 'contract_name'],
  'format': 'json',
  'timezone': 'UTC'},
 'records': [{'datasetid': 'jcdecaux-bike-stations-data-rt',
   'recordid': '6e8b2f6a144a66d65da8d2974bd1a4db0eee4841',
   'fields': {'available_bike_stands': 12,
    'bike_stands': 25,
    'number': 305,
    'address': 'PETER BENOIT - PLACE PETER BENOIT PLAATS 2-4',
    'name': '305 - PETER BENOIT',
    'bonus': 'False',
    'banking': 'False',
    'contract_name': 'Bruxelles-Capitale',
    'status': 'OPEN',
    'available_bikes': 12,
    'position': [50.895592, 4.387965],
    'last_update': '2024-02-26T12:45:24+00:00'},
   'geometry': {'type': 'Point', 'coordinates': [4.387965, 50.895592]}},
  {'datasetid': 'jcdecaux-bike-stations-data-rt',
   'recordid': '39042b9903ec9a87d9fd670c5999fb52ad8d9e6d',
   'fields': {'available_bike_stands': 11,
    'bike_stands': 25,
    'numbe

This is a pretty intricate data structure with dictionaries inside lists inside dictionaries. We can visualize it slightly better by re-encoding it in JSON with indentation.

In [20]:
import json
print(json.dumps(data, indent=4)) # make the output visually more readable

{
    "nhits": 351,
    "parameters": {
        "dataset": "jcdecaux-bike-stations-data-rt",
        "rows": 50,
        "start": 0,
        "facet": [
            "banking",
            "name",
            "bonus",
            "status",
            "contract_name"
        ],
        "format": "json",
        "timezone": "UTC"
    },
    "records": [
        {
            "datasetid": "jcdecaux-bike-stations-data-rt",
            "recordid": "6e8b2f6a144a66d65da8d2974bd1a4db0eee4841",
            "fields": {
                "available_bike_stands": 12,
                "bike_stands": 25,
                "number": 305,
                "address": "PETER BENOIT - PLACE PETER BENOIT PLAATS 2-4",
                "name": "305 - PETER BENOIT",
                "bonus": "False",
                "banking": "False",
                "contract_name": "Bruxelles-Capitale",
                "status": "OPEN",
                "available_bikes": 12,
                "position": [
                    50.895

Another approach is to look at it step by step:

In [21]:
type(data)

dict

Data is a python dictionary => `{key3:value3, key1:value1, key2:value2,...}`

**Dictionaries don't have an order!** but are a set of keys with linked values, where keys can be strings or ints and values can be (nearly) anything (list, dict, string, int, object, ...).

In [22]:
data.keys() # Check what are the keys of the dict

dict_keys(['nhits', 'parameters', 'records', 'facet_groups'])

In [23]:
data['nhits'] # check what is value of key 'nhits'

351

In [24]:
data['parameters'] # check what is value of key 'parameters'

{'dataset': 'jcdecaux-bike-stations-data-rt',
 'rows': 50,
 'start': 0,
 'facet': ['banking', 'name', 'bonus', 'status', 'contract_name'],
 'format': 'json',
 'timezone': 'UTC'}

In [25]:
data['records']# check what is value of key 'records'

[{'datasetid': 'jcdecaux-bike-stations-data-rt',
  'recordid': '6e8b2f6a144a66d65da8d2974bd1a4db0eee4841',
  'fields': {'available_bike_stands': 12,
   'bike_stands': 25,
   'number': 305,
   'address': 'PETER BENOIT - PLACE PETER BENOIT PLAATS 2-4',
   'name': '305 - PETER BENOIT',
   'bonus': 'False',
   'banking': 'False',
   'contract_name': 'Bruxelles-Capitale',
   'status': 'OPEN',
   'available_bikes': 12,
   'position': [50.895592, 4.387965],
   'last_update': '2024-02-26T12:45:24+00:00'},
  'geometry': {'type': 'Point', 'coordinates': [4.387965, 50.895592]}},
 {'datasetid': 'jcdecaux-bike-stations-data-rt',
  'recordid': '39042b9903ec9a87d9fd670c5999fb52ad8d9e6d',
  'fields': {'available_bike_stands': 11,
   'bike_stands': 25,
   'number': 200,
   'address': 'VETERANS COLONIAUX / KOLONIALE VETERANEN - BD JULES GRAINDOR / JULES GRAINDORLAAN',
   'name': '200 - VETERANS COLONIAUX / KOLONIALE VETERANEN',
   'bonus': 'True',
   'banking': 'False',
   'contract_name': 'Bruxelles-Ca

In [26]:
type(data['records']) # how are stored the records ?

list

In [27]:
len(data['records']) # How many records do we have (remember we put &rows=50 in the URL/API call)

50

Let's look at one arbitrary record:

In [28]:
data['records'][7]

{'datasetid': 'jcdecaux-bike-stations-data-rt',
 'recordid': '92ca851251195416b0f7f203cd8a19214e4a55ca',
 'fields': {'available_bike_stands': 16,
  'bike_stands': 25,
  'number': 207,
  'address': 'BIESTEBROECK / BIESTEBROEK - CHEE DE MONS (FACE AU 518-526) / BERGENSESTEENWEG (TEGENOVER 518-526)',
  'name': '207 - BIESTEBROECK / BIESTEBROEK',
  'bonus': 'False',
  'banking': 'False',
  'contract_name': 'Bruxelles-Capitale',
  'status': 'OPEN',
  'available_bikes': 8,
  'position': [50.833424, 4.313378],
  'last_update': '2024-02-26T12:46:36+00:00'},
 'geometry': {'type': 'Point', 'coordinates': [4.313378, 50.833424]}}

In [29]:
#find name of station
print(data['records'][7]['fields']['name'])

207 - BIESTEBROECK / BIESTEBROEK


So the stations info is saved in a list with each element of the list is a dictionary with the info about a particular station

In [30]:
stations = data['records'] # let's save it in a separate variable
stations

[{'datasetid': 'jcdecaux-bike-stations-data-rt',
  'recordid': '6e8b2f6a144a66d65da8d2974bd1a4db0eee4841',
  'fields': {'available_bike_stands': 12,
   'bike_stands': 25,
   'number': 305,
   'address': 'PETER BENOIT - PLACE PETER BENOIT PLAATS 2-4',
   'name': '305 - PETER BENOIT',
   'bonus': 'False',
   'banking': 'False',
   'contract_name': 'Bruxelles-Capitale',
   'status': 'OPEN',
   'available_bikes': 12,
   'position': [50.895592, 4.387965],
   'last_update': '2024-02-26T12:45:24+00:00'},
  'geometry': {'type': 'Point', 'coordinates': [4.387965, 50.895592]}},
 {'datasetid': 'jcdecaux-bike-stations-data-rt',
  'recordid': '39042b9903ec9a87d9fd670c5999fb52ad8d9e6d',
  'fields': {'available_bike_stands': 11,
   'bike_stands': 25,
   'number': 200,
   'address': 'VETERANS COLONIAUX / KOLONIALE VETERANEN - BD JULES GRAINDOR / JULES GRAINDORLAAN',
   'name': '200 - VETERANS COLONIAUX / KOLONIALE VETERANEN',
   'bonus': 'True',
   'banking': 'False',
   'contract_name': 'Bruxelles-Ca

## Stations sorted by name

We can sort the stations by name in alphabetical order by first making a function that for each entry of the list returns the station name. Then we have just to sort the list.

In [65]:
def get_station_name(station):
    "Return the name of a station (station is a dictionary from the API)"
    station_name_raw = station['fields']['name']
    # Remove the station number from the name
    station_name = station_name_raw.split('-', maxsplit = 1)[1].strip()
    return station_name

get_station_name(stations[1])

'VETERANS COLONIAUX / KOLONIALE VETERANEN'

### Loops
To get all the station names we can:
1. Either apply the function to each element of the list manually
2. Or use a loop to do it for us

In [32]:
#1 .  apply the function to each element of the list manually
station_name_list = [get_station_name(stations[0]),
                     get_station_name(stations[1]),
                     get_station_name(stations[2]),
                     get_station_name(stations[3]),
                     get_station_name(stations[4]),
                     get_station_name(stations[5]),
                     get_station_name(stations[6]),
                     get_station_name(stations[7]),
                     get_station_name(stations[8]),
                     # ...
                     get_station_name(stations[49])]
station_name_list

['PETER BENOIT',
 'VETERANS COLONIAUX / KOLONIALE VETERANEN',
 'JOSEPH II / JOZEF II',
 'DUBOIS',
 'JARDIN MASSART',
 'CAMPUS CERIA / CAMPUS VAN CERIA',
 'VANDENDRIESCH',
 'BIESTEBROECK / BIESTEBROEK',
 'FONTAINAS',
 'CROIX DE GUERRES / OORLOGSKRUISEN']

In [33]:
# 2. use a loop to do it for us

station_name_list = [] # initialize empty list
for x in stations: # loop through each element of the list list
    name = get_station_name(x) # get for each station the name
    station_name_list.append(name) # add it to the list

In [34]:
# 2. use a loop to do it for us (loop inside a list = list comprehension)
station_name_list = [get_station_name(x) for x in stations]

In [35]:
# And now just sort it
sorted(station_name_list)

['ADEPS',
 'ARTAN',
 'BASCULE',
 'BASILIQUE / BASILIEK',
 'BERKENDAEL / BERKENDAAL',
 'BIESTEBROECK / BIESTEBROEK',
 'BOILEAU',
 'CAMPUS CERIA / CAMPUS VAN CERIA',
 'CELTES/KELTEN',
 'CENTRE CULTUREL / CULTUREEL CENTRUM',
 'CHAZAL/CHAZAL',
 'CLEMENCEAU',
 'COLONEL BOURG/KOLONEL BOURG',
 'CROIX DE GUERRES / OORLOGSKRUISEN',
 'DARWIN',
 'DOCKS BRUXSEL',
 'DOMAINE / DOMEIN',
 'DUBOIS',
 'EMILE MAX LYCEE/LYCEUM EMILE MAX',
 'FONTAINAS',
 "GARE D'ETTERBEEK",
 'GARE DU NORD/NOORDSTATION',
 'GARE MEISER/MEISER STATION',
 'GEORGES BRUGMANN',
 'GEORGES HENRI',
 'HANKAR',
 'HOUZEAU',
 'JARDIN MASSART',
 'JOSEPH II / JOZEF II',
 'JULES CESAR',
 'KERSBEEK',
 'LES QUAIS / KAAIEN',
 'LOUIS TITZ',
 'LOUISE / LOUIZA',
 'MAISON COMMUNALE DE JETTE GEMEENTEHUIS',
 'MARCEL THIRY',
 'MELBA',
 'MONTGOMERY',
 'PETER BENOIT',
 'PLACE DAILLY/DAILLYPLEIN',
 'PLASKY/PLASKY',
 'PORTE DE FLANDRE / VLAAMSEPOORT',
 'PORTE DE NINOVE / NINOOFSEPOORT',
 'SIBELGA',
 'THEO LAMBERT',
 'THIRY',
 'THOMAS BALIS',
 'UNION SAI

# Make a dictionary with the bike stands names + # of available bikes

In [36]:
station1 = stations[3] #retrieve a station
get_station_name(station1) #name

'DUBOIS'

In [37]:
station1['fields']['available_bikes'] # Number of bikes available

15

In [38]:
dico_number_bikes = {} #initialize empty dictionary
#construct with a for loop
for station in stations:
    dico_number_bikes[get_station_name(station)] = station['fields']['available_bikes']

In [39]:
dico_number_bikes

{'PETER BENOIT': 12,
 'VETERANS COLONIAUX / KOLONIALE VETERANEN': 14,
 'JOSEPH II / JOZEF II': 11,
 'DUBOIS': 15,
 'JARDIN MASSART': 5,
 'CAMPUS CERIA / CAMPUS VAN CERIA': 9,
 'VANDENDRIESCH': 8,
 'BIESTEBROECK / BIESTEBROEK': 8,
 'FONTAINAS': 28,
 'BERKENDAEL / BERKENDAAL': 15,
 'PLACE DAILLY/DAILLYPLEIN': 16,
 'COLONEL BOURG/KOLONEL BOURG': 15,
 'CLEMENCEAU': 22,
 'HANKAR': 9,
 'HOUZEAU': 10,
 'THOMAS BALIS': 9,
 'GARE DU NORD/NOORDSTATION': 22,
 'PORTE DE FLANDRE / VLAAMSEPOORT': 12,
 'GEORGES HENRI': 14,
 'ADEPS': 5,
 'THEO LAMBERT': 10,
 'KERSBEEK': 7,
 "GARE D'ETTERBEEK": 8,
 'LOUIS TITZ': 13,
 'GEORGES BRUGMANN': 14,
 'THIRY': 11,
 'DOMAINE / DOMEIN': 9,
 'ARTAN': 10,
 'MAISON COMMUNALE DE JETTE GEMEENTEHUIS': 12,
 'PLASKY/PLASKY': 15,
 'MELBA': 12,
 'UNION SAINT-GILLOISE / SINT-GILLIS UNIE': 7,
 'LOUISE / LOUIZA': 30,
 'BASCULE': 14,
 'EMILE MAX LYCEE/LYCEUM EMILE MAX': 14,
 'LES QUAIS / KAAIEN': 9,
 'CENTRE CULTUREL / CULTUREEL CENTRUM': 6,
 'DARWIN': 14,
 'BOILEAU': 11,
 'BAS

In [40]:
# other way via dictionary comprehension
dico_number_bikes = {get_station_name(station) : station['fields']['available_bikes'] for station in stations}
dico_number_bikes

{'PETER BENOIT': 12,
 'VETERANS COLONIAUX / KOLONIALE VETERANEN': 14,
 'JOSEPH II / JOZEF II': 11,
 'DUBOIS': 15,
 'JARDIN MASSART': 5,
 'CAMPUS CERIA / CAMPUS VAN CERIA': 9,
 'VANDENDRIESCH': 8,
 'BIESTEBROECK / BIESTEBROEK': 8,
 'FONTAINAS': 28,
 'BERKENDAEL / BERKENDAAL': 15,
 'PLACE DAILLY/DAILLYPLEIN': 16,
 'COLONEL BOURG/KOLONEL BOURG': 15,
 'CLEMENCEAU': 22,
 'HANKAR': 9,
 'HOUZEAU': 10,
 'THOMAS BALIS': 9,
 'GARE DU NORD/NOORDSTATION': 22,
 'PORTE DE FLANDRE / VLAAMSEPOORT': 12,
 'GEORGES HENRI': 14,
 'ADEPS': 5,
 'THEO LAMBERT': 10,
 'KERSBEEK': 7,
 "GARE D'ETTERBEEK": 8,
 'LOUIS TITZ': 13,
 'GEORGES BRUGMANN': 14,
 'THIRY': 11,
 'DOMAINE / DOMEIN': 9,
 'ARTAN': 10,
 'MAISON COMMUNALE DE JETTE GEMEENTEHUIS': 12,
 'PLASKY/PLASKY': 15,
 'MELBA': 12,
 'UNION SAINT-GILLOISE / SINT-GILLIS UNIE': 7,
 'LOUISE / LOUIZA': 30,
 'BASCULE': 14,
 'EMILE MAX LYCEE/LYCEUM EMILE MAX': 14,
 'LES QUAIS / KAAIEN': 9,
 'CENTRE CULTUREL / CULTUREEL CENTRUM': 6,
 'DARWIN': 14,
 'BOILEAU': 11,
 'BAS

## Check the number of stations have at least 20 available bikes

In [41]:
Number = 0 # initialize counter to 0
# loop over a dictionary with .items()
for key,value in dico_number_bikes.items():
    if value >= 20 :
        print('Station ' + key + ' has ' + str(value) + ' bikes available')
        Number += 1 # add 1 to counter (is the same as writing Number = Number + 1)

print('\n') #print blank line
print('In total ' + str(Number) +' stations have at least 20 bikes available')

Station FONTAINAS has 28 bikes available
Station CLEMENCEAU has 22 bikes available
Station GARE DU NORD/NOORDSTATION has 22 bikes available
Station LOUISE / LOUIZA has 30 bikes available


In total 4 stations have at least 20 bikes available


## Exercises
Note: You will have to adapt the `URL_Villo` and make a new request to get all the stations instead of the 50 first ones.

1. How many bike stands are there in total ?
2. How many bikes are available **at this moment** ?
3. Write a function that returns the `stations` list at the **current** time.
3. Write a function that returns the **current** total number of bikes available.
4. Write a function that returns the **current** number of bikes available in an arbitrary station (for instance '351 - ULB').
6. Write a function that checks for each station in `dico_number_bikes` if (in the meanwhile) anyone took or brought back a bike.

### Bonus

1. Write a function that will check every 10 seconds if someone took or brought back a bike in any station during 2 minutes.
2. Find the currently OPEN station that is the least used/ that had the longest time ago an update.

In [42]:
# Redo an API request to get all the bike stations


In [43]:
# Solution
URL_Villo = "https://public.opendatasoft.com/api/records/1.0/search/?dataset=jcdecaux-bike-stations-data-rt&q=&facet=banking&facet=name&exclude.name=BORNE+TEST+ALLEE+VERTE&exclude.name=BORNE+ATELIER+MECANOS+SPL&facet=bonus&facet=status&facet=contract_name&refine.contract_name=Bruxelles-Capitale&rows=1000"
response = requests.get(URL_Villo)  # Retrieve the response
data = response.json()  # Parse from JSON

    1. How many bike stands are there in total ?

In [45]:
# Solution
len(data['records'])

351

    2. How many bikes are available at this moment ?

*Hint: to answer this question, you need to have the latest info available (otherwise it is not "at this moment") => make a new API call and compute answer directly after it)*

In [46]:
# Solution
response = requests.get(URL_Villo)  # Retrieve the response
data = response.json()  # Parse from JSON
stations = data['records']
total = 0
for stat in stations:
    total = total + stat['fields']['available_bikes']

In [47]:
total

3807

    3. Write a function that returns the stations list at the current time.

In [48]:
# Solution
def get_villo_stations():
    "Get new data from the Villo API, for the given Villo API URL."

    response = requests.get(URL_Villo)  # Retrieve the response
    data = response.json()  # Parse from JSON
    stations = data['records']
    return stations

#test:
stations = get_villo_stations()
print(len(stations))

351


    4. Write a function that returns the current total number of bikes available.
*Hint: Same as before but write it as a function*

In [49]:
# Solution
def current_available_bikes():
    stations = get_villo_stations()
    total_bikes_available = 0 #initialize counter at 0

    for station in stations:
        # Loop through all stations and add the number of availaible bikes to counter
        total_bikes_available += station['fields']['available_bikes']

    return(total_bikes_available)

#test
current_available_bikes()

3807

    5. Write a function that returns the current number of bikes available in an arbitrary station (for instance 'ULB').

In [50]:
# Solution
def bikes_at_station(stations, station_to_find = 'ULB'):

    for station in stations:
        if get_station_name(station) == station_to_find:
            # This is the station we are looking for
            return station['fields']['available_bikes']


In [51]:
bikes_at_station(get_villo_stations(),'ULB')

20

In [52]:
bikes_at_station(stations = get_villo_stations())

20

    6. Check for each station in `dico_number_bikes` if (in the meanwhile) anyone took or brought back a bike.

In [53]:
#Solution
stations = get_villo_stations()
for station, bikes in dico_number_bikes.items():
    now = bikes_at_station(stations,station)
    if bikes != now:
        print(f'For station {station} there were {bikes} bikes and now there are {now} bikes')

**Bonus**

    1. Write a function that will check every 10 seconds if someone took or brought back a bike in any station during 2 minutes.
*Hint: use python package time and the functionality time.sleep(10) to wait 10 seconds between each call and use package datetime and function datetime.datetime.now() to print the time stamp at each step*

In [63]:
#Solution
import time, datetime
def Check_changes_10s():
    stations = get_villo_stations()
    dico_number_bikes_init = {get_station_name(station) : station['fields']['available_bikes'] for station in stations}
    print('Start at: ' + str(datetime.datetime.now()))
    for i in range(12): # do 12 times (10 sec * 12 = 120 sec = 2 min)
        time.sleep(10)
        print(f'Loop {i} at {str(datetime.datetime.now())}')
        stations = get_villo_stations()
        dico_number_bikes_new = {get_station_name(station) : station['fields']['available_bikes'] for station in stations}
        for station, bikes in dico_number_bikes_new.items():
            if bikes != dico_number_bikes_init[station]:
                print(f'For station {station} there were {dico_number_bikes_init[station]} bikes and now there are {bikes} bikes')
        dico_number_bikes_init = dico_number_bikes_new

In [ ]:
Check_changes_10s()

2. Find the currently OPEN station that is the least used/ that had the longest time ago an update.

In [57]:
# Solution
def find_least_used(stations = get_villo_stations()):
    # initialize with time in future
    last_update_time = '2200-01-01T00:00:00+00:00'
    for station in stations:  # loop over stations
        if station['fields']['last_update'] < last_update_time and station['fields']['status'] == 'OPEN': # check update time smaller and still open
            last_updated = station
            last_update_time = station['fields']['last_update']
    print(f'Station {get_station_name(last_updated)} was last used at time {last_update_time}')
    return last_updated

In [58]:
find_least_used()

Station AUBERGE DE JEUNESSE was last used at time 2023-12-04T10:06:45+00:00


{'datasetid': 'jcdecaux-bike-stations-data-rt',
 'recordid': 'e4e47ef9669d7bdb54b89b98ffabfe35f31ef76f',
 'fields': {'available_bike_stands': 6,
  'bike_stands': 25,
  'number': 192,
  'address': 'AUBERGE DE JEUNESSE - RUE DELAUNOY/STRAAT ,67',
  'name': '192 - AUBERGE DE JEUNESSE',
  'bonus': 'False',
  'banking': 'False',
  'contract_name': 'Bruxelles-Capitale',
  'status': 'OPEN',
  'available_bikes': 19,
  'position': [50.851562, 4.332924],
  'last_update': '2023-12-04T10:06:45+00:00'},
 'geometry': {'type': 'Point', 'coordinates': [4.332924, 50.851562]}}